src/test/java/org/jenkinsci/plugins/pipeline/milestone/MilestoneStepTest.java

```diff
@@ -0,0 +1,209 @@
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 passed milestone 1
+
+                // Let #2 continue so it goes away from milestone 1 (and passes milestone 2)
+                SemaphoreStep.success("wait/2", null);
+                story.j.waitForCompletion(b2);
+
+                // Once #2 continues and passes milestone 2 then #1 is automatically cancelled
+                story.j.assertBuildStatus(Result.NOT_BUILT, story.j.waitForCompletion(b1));
+            }
+        });
+    }
+
+    @Test
+    public void olderBuildsMustBeCancelledOnBuildFinish() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "milestone label: 'My Label'\n" +
+                        "node {\n" +
+                        "  echo 'First milestone'\n" +
+                        "  semaphore 'wait'\n" +
+                        "}"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 passed milestone 1
+
+                // Let #2 continue
+                SemaphoreStep.success("wait/2", null);
+                story.j.waitForCompletion(b2);
+
+                // Once #2 finishes, so it passes the virtual ad-inifinitum milestone, then #1 is automatically cancelled
+                story.j.assertBuildStatus(Result.NOT_BUILT, story.j.waitForCompletion(b1));
+            }
+        });
+    }
+
+    @Test
+    public void olderBuildsMustNotBeCancelledOnBuildAborted() {
+        story.addStep(new Statement() {
+            @Override public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+                p.setDefinition(new CpsFlowDefinition(
+                        "milestone label: 'My Label'\n" +
+                        "node {\n" +
+                        "  echo 'First milestone'\n" +
+                        "  semaphore 'wait'\n" +
+                        "}"));
+                WorkflowRun b1 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/1", b1);
+                WorkflowRun b2 = p.scheduleBuild2(0).waitForStart();
+                SemaphoreStep.waitForStart("wait/2", b2);
+                // Now both #1 and #2 passed milestone 1
+
+                // Abort #2
+                b2.getOneOffExecutor().doStop();
+                story.j.waitForMessage("Finished: ABORTED", b2);
+
+                assertTrue(b1.isBuilding()); // #1 shuould not be cancelled
```

Comment:
I suppose your thinking is that `b1` had some changes eligible for “production” or whatever, and `b2` had some newer ones, but when the user decided to kill off `b2` for whatever reason, `b1` is still valid. Perhaps; but you are only detecting that the status is `ABORTED`, not _why_ it has that status. You are certainly not verifying that it was _manually_ aborted.

What if `b2` had a major rewrite of the code that renders `b1` obsolete, but the `Jenkinsfile` happened to have a too-short `timeout` to deal with all the new test cases, leading to `b2` becoming aborted on build timeout? Surely this is not conceptually different from `b2` ending with `FAILURE` due to a broken test run by some test runner that just returns a nonzero exit code, or equivalently `b2` being marked `UNSTABLE` by `JUnitResultArchiver` and the script then `return`ing prior to a production stage based on that test result.

In any of these cases, `b2` did not go all the way through, so someone will have to correct the problem and trigger `b3`. We do not know for sure whether it would make sense to continue with `b1`. If you want to err on the side of preserving `b1`, then I think you just have to abandon the virtual terminal milestone. Perhaps nothing is lost by that simplification: assuming you have an explicit milestone before any “go to production” steps, you are still ensuring that you do not clobber newer bits with older bits, right?